test_util/test_upgrade_vpc.py

```diff
@@ -159,125 +132,279 @@ def wait_for_dns(hostname):
-    # This is a hack to make sure we don't deploy dns_app before the name it's
-    # trying to resolve is available.
-    wait_for_dns(dns_app['env']['RESOLVE_NAME'])
-    cluster_api.marathon.deploy_app(dns_app, check_health=False)
-    cluster_api.marathon.ensure_deployments_complete()
-
-    test_apps = [healthcheck_app, dns_app]
-    test_app_ids = [app['id'] for app in test_apps]
-
-    tasks_start = {app_id: sorted(app_task_ids(app_id)) for app_id in test_app_ids}
-    log.debug('Test app tasks at start:\n' + pprint.pformat(tasks_start))
-
-    for app in test_apps:
-        assert app['instances'] == len(tasks_start[app['id']])
-
-    yield
-
-    if not cluster_api.marathon.get('/v2/apps').json()['apps']:
-        raise Exception('No Marathon apps running!')
-
-    tasks_end = {app_id: sorted(app_task_ids(app_id)) for app_id in test_app_ids}
-    log.debug('Test app tasks at end:\n' + pprint.pformat(tasks_end))
-
-    # Verify that the tasks we started are still running.
-    assert tasks_start == tasks_end
-
-    # Verify DNS didn't fail.
-    marathon_framework_id = cluster_api.marathon.get('/v2/info').json()['frameworkId']
-    dns_app_task = cluster_api.marathon.get('/v2/apps' + dns_app['id'] + '/tasks').json()['tasks'][0]
-    dns_log = parse_dns_log(cluster_api.mesos_sandbox_file(
-        dns_app_task['slaveId'],
-        marathon_framework_id,
-        dns_app_task['id'],
-        dns_app['env']['DNS_LOG_FILENAME'],
-    ))
-    dns_failure_times = [entry[0] for entry in dns_log if entry[1] != 'SUCCESS']
-    if len(dns_failure_times) > 0:
-        message = 'Failed to resolve Marathon app hostname {} at least once.'.format(dns_app['env']['RESOLVE_NAME'])
-        log.debug(message + ' Hostname failed to resolve at these times:\n' + '\n'.join(dns_failure_times))
-        # Skip raising an exception on DNS failure so that other tests can run.
-        # DNS failure is being tracked at
-        # https://mesosphere.atlassian.net/browse/DCOS-13426.
-        #raise Exception(message)  # noqa
+
+class VpcClusterUpgradeTestDcosApiSessionFactory:
+    def apply(self, dcos_url: str, masters: List[str], public_masters: List[str], slaves: List[str],
+              public_slaves: List[str], default_os_user: str) -> DcosApiSession:
+        pass
+
+
+class VpcClusterUpgradeTest:
+    log = logging.getLogger(__name__)
+
+    def __init__(self,
+                 num_masters: int, num_agents: int, num_public_agents: int,
+                 stable_installer_url: str, installer_url: str,
+                 aws_region: str, aws_access_key_id: str, aws_secret_access_key: str,
+                 ssh_key: str, default_os_user: str,
+                 config_yaml_override_install: str, config_yaml_override_upgrade: str,
+                 dcos_api_session_factory_install: VpcClusterUpgradeTestDcosApiSessionFactory,
+                 dcos_api_session_factory_upgrade: VpcClusterUpgradeTestDcosApiSessionFactory):
```

Comment:
What would the use sight look like then? I don't want to loose the "Give me a thing that when I invoke it will give me a `DcosApiSession`" that I get with the factory class right now (admittedly this is my python ignorance) the whole point of introducing the factory is that user is able to provide the path, but the test controls the time when creating the client.